text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,306 @@
+
+#### Other constraints
+
+The directives may appear in any order on the element.
+
+Any other constraints that apply to `for:each` must also apply to `lwc:each`. For example, `lwc:ref` is (currently) unsupported inside of `for:each` and `iterator:*` – the same applies to `lwc:each`. For another example, `<slot>`s cannot have any of the new directives, and cannot be repeated inside the new iterator.
+
+## Drawbacks
+
+This new directive may cause confusion, because we are introducing a third directive to try to unify the previous two (see [XKCD 927](https://xkcd.com/927/)).
+
+The implementation also becomes more complex, as we will need to support all three directives for the foreseeable future. Perhaps they can share logic under the hood, but we will still need to test all three, at the very least.
+
+We will also need to make an effort to advocate for adoption of the new directive, which may involve updating existing documentation and writing codemods or IDE tooling.
+
+## Alternatives
+
+### Enhancing `for:each`
+
+One reasonable alternative would be to enhance the existing `for:each` rather than creating a new directive. However, this has several drawbacks:
+
+1. There could be conflicts between the legacy `key` on within-iterator elements versus the `lwc:key` on the iterator element. In cases of nested iterators, it could be very tricky to determine which keys should apply to which iterators.
+2. It does not move us closer to a world where all LWC directives use the `lwc:` prefix. Directives like `for:each`, `for:item`, `for:index`, and `key` would continue to be the way to write iterators, making it difficult to support a [directive shorthand](https://github.com/salesforce/lwc/issues/3303) in the future.
+
+### Enhancing `iterator:*`
+
+Similarly, enhancing `iterator:*` is another approach. However, this directive is seldom-used, and is rigid in how it allows for referencing the value, index, and first and last items in a list (the property names must always be `value`, `index`, `first`, and `last`).
+
+### Making `lwc:key` optional
+
+The key is a performance optimization, and not all developers need it. Some developers even use a getter with `Math.random()` to provide a random key, to work around the current LWC requirement that lists must have a `key`.
+
+It is entirely possible to make `lwc:key` optional. However, this comes with a few downsides:
+
+1. Developers may be encouraged to use a less-performant rendering pattern, since it takes less effort to omit the key.
+2. It deviates from the existing behavior of `for:each`, and is one more thing to teach.
+3. Developers may now assume that keys are actually a _bad_ idea, since the "improved" `lwc:each` makes them optional.
+
+So, for this proposal, keys are still required. However, it is definitely something that can be revisited in the future.

```

Comment:
@pmdartus: Use case for making it optional:

- You have a list of tags, it's just a list of strings, they'll never change
- You use the tag itself as the key since the key is required
- Now imagine suddenly tags are duplicated, now you need a unique id for each tag
- Now you have to add special IDs or `Math.random()` to work around it

Reply:
@ekashida: We could add another directive like `lwc:no-key` to communicate that a key is not required.

Reply:
Or `lwc:dangerously-set-no-key-this-will-be-slow-and-i-like-it-that-way`

Reply:
I looked into how other frameworks do this:

- Svelte/Vue/Angular: keys are optional (Angular calls it `TrackByFunction` rather than "key")
- React: keys are not required, but you will get a [console warning](https://react.dev/learn/rendering-lists#rendering-data-from-arrays) without them
- Lit: keys are effectively optional: you can use `Array.prototype.map` for the key-less approach, or [the `repeat` directive](https://lit.dev/docs/templates/lists/#the-repeat-directive) which requires a `keyFunction`
- Solid: More subtle. `<For>` does not require an explicit key function [by design](https://github.com/solidjs/solid/discussions/366#discussioncomment-459799) since Solid already has fine-grained reactivity. However it is keyed based on referential equality. Whereas there is the alternative [`<Index>` component](https://docs.solidjs.com/references/api-reference/control-flow/Index) which is non-keyed and designed for cases where referential equality doesn't make sense (such as primitives).